File: 剑指offer/前缀树/OfferII063.py
class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        dictionarySet = set(dictionary)
        words = sentence.split()
        for i, word in enumerate(words):
            for j in range(1, len(word)):
                if word[:j] in dictionarySet:
                    words[i] = word[:j]
                    break
        return " ".join(words)

def mian():
    solution = Solution()
    dictionary = ['cat', 'bat', 'rat']
    sentence = "the cattle was rattled by the battery"
    result = solution.replaceWords(dictionary, sentence)
    print(result)


if __name__ == "__main__":
    main()
    

File: 剑指offer/前缀树/OfferII063.cpp
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {
        unordered_set<string> dict(dictionary.begin(), dictionary.end());
        vector<string> words;
        string word = "";
        for (char c : sentence) {
            if (c == ' ') {
                words.push_back(word);
                word = "";
            } else {
                word += c;
            }
        }
        words.push_back(word);
        for (int i = 0; i < words.size(); i++) {
            string& word = words[i];
            for (int j = 1; j <= word.length(); j++) {
                if (dict.count(word.substr(0, j))) {
                    word = word.substr(0, j);
                    break;
                }
            }
        }
        string result = "";
        for (int i = 0; i < words.size(); i++) {
            result += words[i];
            if (i < words.size() - 1) {
                result += " ";
            }
        }
        return result;
    }
};
